### PR TITLE
New code flow queue

### PR DESCRIPTION
**Pull request by dkurepa**
https://github.com/dotnet/arcade-services/issues/4061

All reminders are still configured to use the default work item queue. I don't think the codeflow subscriptions create any reminders currently, @premun please correct me if I'm wrong here

### CODE REVIEW

**Comment by premun** [comment]
The codeflow reminders need to also go in the VMR queue because they are basically also subscription reminders. This means that the processors for them will continue interacting with VMR subscriptions so they also need to go in a separate queue

**Comment by dkurepa** [comment]
> The codeflow reminders need to also go in the VMR queue because they are basically also subscription reminders. This means that the processors for them will continue interacting with VMR subscriptions so they also need to go in a separate queue

Yeah I understand that they do, I was checking if they exist, sounds like they do. I'll modify the code to reflect that

File: src/ProductConstructionService/ProductConstructionService.SubscriptionTriggerer/SubscriptionTriggerer.cs [hunk 1]
```diff
@@ -14,36 +15,49 @@ public class SubscriptionTriggerer
...
+        _defaultWorkItemProducerFactory = defaultWorkItemProducerFactory;
+        _codeflowWorkItemProducerFactory = codeflowWorkItemProducerFactory;
         _subscriptionIdGenerator = subscriptionIdGenerator;
     }
 
     public async Task TriggerSubscriptionsAsync(UpdateFrequency targetUpdateFrequency)
     {
-        var workItemProducer = _workItemProducerFactory.CreateProducer<SubscriptionTriggerWorkItem>();
+        var defaultWorkItemProducer = _defaultWorkItemProducerFactory.CreateProducer<SubscriptionTriggerWorkItem>();
+        var codeflowWorkItemProducer = _codeflowWorkItemProducerFactory.CreateProducer<SubscriptionTriggerWorkItem>();
```
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.SubscriptionTriggerer/SubscriptionTriggerer.cs:39]: I wonder how much easier would it be to just return the right producer from the factory. Technically the `if/else` would be implemented just once there while we can have any number of places producing these work items all around the codebase which now would have to have the `if/else` every time

Comment (dkurepa) [reply src/ProductConstructionService/ProductConstructionService.SubscriptionTriggerer/SubscriptionTriggerer.cs:39]: The factory is already configured to a queue, so it wouldn't work in the current implementation, we'd need a `FactoryFactory` 😅

Comment (dkurepa) [reply src/ProductConstructionService/ProductConstructionService.SubscriptionTriggerer/SubscriptionTriggerer.cs:39]: oh or no, we just pass the queueName to the `IWorkItemProducer`

Comment (premun) [reply src/ProductConstructionService/ProductConstructionService.SubscriptionTriggerer/SubscriptionTriggerer.cs:39]: Yeah, I don't see why it wouldn't know about both and accept a parameter `isCodeFlowSubscription` or something. We wouldn't have to inject two factories everywhere then

File: src/ProductConstructionService/ProductConstructionService.DependencyFlow/NonBatchedPullRequestUpdater.cs [hunk 2]
```diff
@@ -66,8 +66,12 @@ public NonBatchedPullRequestUpdater(
         Subscription? subscription = await _context.Subscriptions.FindAsync(SubscriptionId);
         if (subscription == null)
         {
-            await _pullRequestCheckReminders.UnsetReminderAsync();
-            await _pullRequestUpdateReminders.UnsetReminderAsync();
+            // We don't know if the subscription was a code flow one, so just uncheck both
```
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.DependencyFlow/NonBatchedPullRequestUpdater.cs:69]: We do know - `subscription.SourceEnabled`

Comment (premun) [reply src/ProductConstructionService/ProductConstructionService.DependencyFlow/NonBatchedPullRequestUpdater.cs:69]: Technically this is maybe something we could bind at the manager creation time and not have to supply the boolean?

File: src/ProductConstructionService/ProductConstructionService.Api/Api/v2020_02_20/Controllers/BuildsController.cs [hunk 3]
```diff
@@ -324,7 +324,8 @@ await _context.BuildDependencies.AddRangeAsync(
 
         // Compute the dependency incoherencies of the build.
         // Since this might be an expensive operation we do it asynchronously.
-        await _workItemProducerFactory.CreateProducer<BuildCoherencyInfoWorkItem>()
+        var a = _workItemProducerFactory.CreateProducer<BuildCoherencyInfoWorkItem>();
+        await a
```
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.Api/Api/v2020_02_20/Controllers/BuildsController.cs]: I guess this variable was used for testing?

File: test/ProductConstructionService.DependencyFlow.Tests/PullRequestUpdaterTests.cs [hunk 4]
```diff
@@ -455,6 +455,7 @@ protected void AndShouldHavePullRequestCheckReminder()
         {
             UpdaterId = GetPullRequestUpdaterId().ToString(),
             Url = prUrl,
+            IsCodeFlow = isCodeFlow
```
Comment (premun) [inline test/ProductConstructionService.DependencyFlow.Tests/PullRequestUpdaterTests.cs:458]: We can get this from like in line 450

File: src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemProducerFactory.cs [hunk 5]
```diff
@@ -7,14 +7,17 @@ namespace ProductConstructionService.WorkItems;
...
-    private readonly string _queueName = queueName;
+    private readonly string _defaultQueueName = defaultQueueName;
+    private readonly string _codeFlowQueueName = codeFlowQueueName;
 
-    public IWorkItemProducer<T> CreateProducer<T>() where T : WorkItem
-        => new WorkItemProducer<T>(_queueServiceClient, _queueName);
+    public IWorkItemProducer<T> CreateProducer<T>(bool IsCodeFlowSubscription = false) where T : WorkItem
+        => IsCodeFlowSubscription
+            ? new WorkItemProducer<T>(_queueServiceClient, _codeFlowQueueName)
+            : new WorkItemProducer<T>(_queueServiceClient, _defaultQueueName);
```
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemProducerFactory.cs:22]:
```suggestion
    public IWorkItemProducer<T> CreateProducer<T>(bool isCodeFlowSubscription = false) where T : WorkItem
        => isCodeFlowSubscription
            ? new WorkItemProducer<T>(_queueServiceClient, _codeFlowQueueName)
            : new WorkItemProducer<T>(_queueServiceClient, _defaultQueueName);
```

File: test/ProductConstructionService.DependencyFlow.Tests/PullRequestUpdaterTests.cs (hunk 4 above)
Comment (dkurepa) [reply test/ProductConstructionService.DependencyFlow.Tests/PullRequestUpdaterTests.cs]: yeah, didn't notice it there fixed

File: test/ProductConstructionService.DependencyFlow.Tests/UpdateAssetsTests.cs [hunk 6]
```diff
@@ -34,7 +34,7 @@ public async Task UpdateWithAssetsNoExistingPR(bool batchable)
         AndCreateNewBranchShouldHaveBeenCalled();
         AndCommitUpdatesShouldHaveBeenCalled(b);
         AndCreatePullRequestShouldHaveBeenCalled();
-        AndShouldHavePullRequestCheckReminder();
+        AndShouldHavePullRequestCheckReminder(false);
```
[the remark this answers is not in the document]
Comment (dkurepa) [reply test/ProductConstructionService.DependencyFlow.Tests/UpdateAssetsTests.cs]: true, I was looking at the test file name and putting corresponding boolean there. It's not needed now tho

**Review by premun — APPROVED** [review]

**Pull request merged**